AlphaWallet/Core/CoinTicker/CoinTickersFetcher.swift

```diff
@@ -0,0 +1,310 @@
+        self.config = config
+    }
+
+    private func fetchAllCoinGeckoTickers() -> Promise<[Ticker]> {
+        if CoinTickersFetcher.coinGeckoTickers.isEmpty {
+            return firstly {
+                provider.request(.coins(config: config))
+            }.map(on: CoinTickersFetcher.queue, { response -> [Ticker] in
+                return try response.map([Ticker].self, using: JSONDecoder())
+            }).map(on: CoinTickersFetcher.queue, { tickers in
+                CoinTickersFetcher.coinGeckoTickers = tickers
+
+                return tickers
+            }).recover { _ -> Promise<[Ticker]> in
+                return .value([])
+            }
+        } else {
+            return .value(CoinTickersFetcher.coinGeckoTickers)
+        }
+    }
+
+    private var isFetchingTickersAvailable: Bool {
+        switch server {
+        case .main, .xDai:
+            return true
+        case .kovan, .ropsten, .rinkeby, .poa, .sokol, .classic, .callisto, .goerli, .artis_sigma1, .artis_tau1, .binance_smart_chain, .binance_smart_chain_testnet, .custom, .heco, .heco_testnet, .fantom, .fantom_testnet, .avalanche, .avalanche_testnet, .polygon, .mumbai_testnet:
+            return false
+        }
+    }
+
+    private struct MappedCoinTickerId: Hashable {
+        let coinTickerId: String
+        let contractAddress: AlphaWallet.Address
+        let server: RPCServer
+
+        init(id: String, contractAddress: AlphaWallet.Address, server: RPCServer) {
+            self.coinTickerId = id
+            self.contractAddress = contractAddress
+            self.server = server
+        }
+    }
+
+    func fetchCoinTickers(for enabledObjects: [TokenMappedToTicker]) {
+        fetchCoinTickersPromise(for: enabledObjects).done { [weak self] tickers in
+            self?.tickers = tickers
+
+            if !tickers.isEmpty {
+                self?.fetchChartHistoryForAlearyResolvedTickers()
+            }
+        }.catch { _ in
+            //no-op
+        }
+    }
+
+    private static let fetchChartHistoryTimeLimit: TimeInterval = 5
+    private let loadChartHistoryLimiter = TimedLimiter(limit: CoinTickersFetcher.fetchChartHistoryTimeLimit)
+
+    func fetchChartHistoryForAlearyResolvedTickers() {
+        loadChartHistoryLimiter.execute { [weak self] in
+            guard let strongSelf = self else { return }
+            guard let value = strongSelf.tickers.first(where: { strongSelf.history[$0.value] == nil }) else { return }
+
+            strongSelf.loadAllPeriodsHistory(for: value.key).done { _ in
+                CoinTickersFetcher.queue.asyncAfter(deadline: .now() + Self.fetchChartHistoryTimeLimit + 1) {
+                    strongSelf.fetchChartHistoryForAlearyResolvedTickers()
+                }
+            }.cauterize()
+        }
+    }
+
+    private func loadAllPeriodsHistory(for key: AddressToPRCServerKey) -> Promise<Void> {
+        return Promise<Void> { seal in
+            let group = DispatchGroup()
```

Comment:
If we keep this, can we abstract this line (and related lines) that are more management-related plumbing, then logic?

```
let group = DispatchGroup()
```

Reply:
Maybe we can use `when(resolved:)` here instead too. Higher level.